apiserver/introspection.go

```diff
@@ -42,8 +47,32 @@ func (h introspectionHandler) checkAuth(r *http.Request) error {
 	// or "read" access on the controller model, can
 	// access these endpoints.
 
+	accessService := h.ctx.srv.shared.serviceFactoryGetter.FactoryForModel(database.ControllerNS).Access()
+
+	userPermission := func(subject names.UserTag, target names.Tag) (permission.Access, error) {
+		var pID permission.ID
+		switch target.Kind() {
+		case names.ModelTagKind:
+			pID = permission.ID{
+				ObjectType: permission.Model,
+				Key:        target.Id(),
+			}
+		case names.ControllerTagKind:
+			pID = permission.ID{
+				ObjectType: permission.Model,
+				// TODO (manadart 2024-06-28): This needs to work on the controller UUID.
+				Key: "controller",
+			}
+		default:
+			return "", errors.NotValidf("%q as a target", target.Kind())
+		}
+
+		access, err := accessService.ReadUserAccessForTarget(context.TODO(), subject.Id(), pID)
```

Comment:
Should this access service be behind an interface for testing? Or is that not done for the API server

Reply:
All the API server login/auth tests manipulate the DB directly.